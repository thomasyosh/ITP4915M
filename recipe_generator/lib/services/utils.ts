import { clsx, type ClassValue } from "clsx"
import { twMerge } from "tailwind-merge"

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs))
}

export function getStrapiURL() {
  return process.env.STRAPI_API_URL ?? "https://cms.ite4108m.com";
}