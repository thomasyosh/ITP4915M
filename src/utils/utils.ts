import axios from "axios";
import { getSession } from 'next-auth/react';
import { ExtendedRecipe, PaginationQueryType } from "../types";
import { GetServerSidePropsContext } from "next";

// Filters the results by enhancing recipe information with ownership and liked status for the user
export const filterResults = (recipes: ExtendedRecipe[], userId: string) => {
  return recipes.map((recipe) => (
    {
      ...recipe,
      owner: {
        _id: recipe.owner._id,
        name: recipe.owner.name,
        image: recipe.owner.image
      },
      likedBy: recipe.likedBy.map(({ _id, name, image }) => ({ _id, name, image })), // Simplifies likedBy list
      owns: recipe.owner._id.toString() === userId, // Flags if the recipe belongs to the user
      liked: recipe.likedBy.some(l => l._id.toString() === userId) // Flags if the user liked the recipe
    }
  ))
}

// Updates the recipe list by either replacing or removing a recipe from the list
export const updateRecipeList = (
  oldList: ExtendedRecipe[],
  newRecipe: ExtendedRecipe | null,
  deleteId?: string
) => {
  if (!newRecipe && !deleteId) return oldList
  const id = newRecipe ? newRecipe._id : deleteId;
  return newRecipe
    ? oldList.map(recipe => (recipe._id === id ? newRecipe : recipe))
    : oldList.filter(recipe => recipe._id !== id);
};

// Utility to fetch data on server-side while ensuring user authentication
export const getServerSidePropsUtility = async (context: GetServerSidePropsContext, address: string, propskey: string = 'recipes') => {
  try {
    const session = await getSession(context);
    if (!session) {
      return {
        redirect: {
          destination: '/',
          permanent: false,
        },
      };
    }
    const { data } = await axios.get(`${process.env.NEXT_PUBLIC_API_BASE_URL}/${address}`, {
      headers: {
        Cookie: context.req.headers.cookie || '',
      },
    });
    return {
      props: {
        [propskey]: data,
      },
    };
  } catch (error) {
    console.error(`Failed to fetch ${propskey}:`, error); // Logs errors in fetching data
    return {
      props: {
        [propskey]: [], // Returns an empty list if there's an error
      },
    };
  }
};

// REST API call utility supporting multiple HTTP methods
interface methods {
  put: 'put';
  post: 'post';
  delete: 'delete';
  get: 'get';
}

interface RESTcallTypes {
  address: string;
  method?: keyof methods;
  payload?: {
    [key: string]: any;
  };
}

export const call_api = async ({ address, method = 'get', payload }: RESTcallTypes) => {
  try {
    const { data } = await axios[method as keyof methods](address, payload);
    return data; // Returns the data from the API call
  } catch (error) {
    console.error(`An error occurred making a ${method} REST call to -> ${address} error -> ${error}`);
    throw (error); // Rethrows the error for further handling
  }
};

export const formatDate = (date: string) => {
  const [, day, mth, year] = new Date(date).toUTCString().split(' ');
  return `${day} ${mth} ${year}`;
};

export const playAudio = async (
  audioUrl: string,
  audioRef: React.MutableRefObject<HTMLAudioElement | null>,
  onEnd?: () => void // Optional callback when audio finishes
): Promise<void> => {
  try {
    const audio = new Audio(audioUrl);
    audio.preload = 'auto'; // Force preloading
    audioRef.current = audio; // Save the audio instance

    // Attach the `ended` event listener
    audio.onended = () => {
      if (onEnd) onEnd(); // Call the callback if provided
    };

    // Explicitly force loading
    audio.load();

    // Wait for the audio to preload
    await new Promise<void>((resolve, reject) => {
      let isResolved = false;

      audio.oncanplaythrough = () => {
        if (!isResolved) {
          isResolved = true;
          resolve();
        }
      };

      audio.onerror = () => {
        if (!isResolved) {
          isResolved = true;
          reject(new Error('Error loading audio'));
        }
      };

      setTimeout(() => {
        if (!isResolved) {
          isResolved = true;
          reject(new Error('Audio loading timeout'));
        }
      }, 20000); // 20 seconds timeout
    });

    // Attempt playback
    await audio.play();
  } catch (error: any) {
    console.error(`playAudio: Error playing audio: ${error.message}`);
    throw error;
  }
};

export const paginationQueryHelper = (queryObj: PaginationQueryType) => {
  const page = Number(queryObj.page) || 1;
  const limit = Number(queryObj.limit) || 12; // Default: 12 recipes per page
  const skip = (page - 1) * limit;
  const sortOption = typeof queryObj.sortOption === 'string' ? queryObj.sortOption : 'popular';
  const query = typeof queryObj.query === 'string' ? queryObj.query : undefined;

  return { page, limit, skip, sortOption, query };
};