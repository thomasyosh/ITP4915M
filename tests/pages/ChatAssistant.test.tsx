import ChatAssistantPage from "../../src/pages/ChatAssistant";
import { render, screen } from '@testing-library/react';
import { useRouter } from 'next/router';
import * as apiCalls from "../../src/utils/utils";
import { stub_recipe_1 } from '../stub';

const routePushMock = jest.fn();

jest.mock("next/router", () => ({
    useRouter: jest.fn(() => ({
        pathName: 'mocked Path',
        push: routePushMock,
        query: {
            recipeId: 'stub_recipeid'
        },
        events: {
            on: jest.fn(),
            off: jest.fn()
        }
    })),
}));

jest.mock("next/image");

jest.mock("../../src/utils/utils", () => ({
    ...jest.requireActual("../../src/utils/utils"),
    call_api: jest.fn()
}));

jest.mock('react-markdown', () => {
    const ReactMarkdownMock = ({ children }: { children: React.ReactNode }) => <div>{children}</div>;
    ReactMarkdownMock.displayName = 'ReactMarkdownMock';
    return ReactMarkdownMock;
});

describe('The ChatAssistant Page', () => {
    it('shall successfully render a recipe', async () => {
        const getRecipeFromAPI = jest.spyOn(apiCalls, 'call_api');
        getRecipeFromAPI.mockImplementationOnce(() => Promise.resolve(stub_recipe_1));

        const { container } = render(<ChatAssistantPage />);

        expect(await screen.findByText('Recipe_1_name')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('shall render the error page if no recipe id is provided', async () => {
        (useRouter as jest.Mock).mockImplementationOnce(() => ({
            query: {},
        }));

        render(<ChatAssistantPage />);

        expect(await screen.findByText('Missing recipe ID.')).toBeInTheDocument();
    });

    it('shall render the error page if there is an API fetch error', async () => {
        const getRecipeFromAPI = jest.spyOn(apiCalls, 'call_api');
        getRecipeFromAPI.mockImplementationOnce(() => Promise.reject({ message: 'Fetch error message' }));

        render(<ChatAssistantPage />);

        expect(await screen.findByText('Fetch error message')).toBeInTheDocument();
    });

    it('shall render the error page if no data is returned from API', async () => {
        const getRecipeFromAPI = jest.spyOn(apiCalls, 'call_api');
        getRecipeFromAPI.mockImplementationOnce(() => Promise.resolve(undefined));

        render(<ChatAssistantPage />);

        expect(await screen.findByText('No recipe data found.')).toBeInTheDocument();
    });
});
