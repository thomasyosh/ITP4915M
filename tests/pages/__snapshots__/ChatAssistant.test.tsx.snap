// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ChatAssistant Page shall successfully render a recipe 1`] = `
<div>
  <div
    class="min-h-screen bg-gray-100 p-4"
  >
    <div
      class="max-w-4xl mx-auto bg-white shadow-lg rounded-lg overflow-hidden"
    >
      <div
        class="relative w-full h-80"
      />
      <div
        class="pl-6 pt-6 pr-6"
      >
        <h2
          class="text-2xl font-bold mb-2 mt-2"
        >
          Recipe_1_name
        </h2>
         
        <div
          class="flex items-center mb-6"
        >
          <div
            class="relative w-12 h-12 rounded-full overflow-hidden mr-4"
          />
          <div>
            <span
              class="text-gray-700 text-lg"
            >
              By 
              <span
                class="text-gray-600 hover:text-gray-800 hover:underline cursor-pointer"
              >
                user_1
              </span>
            </span>
            <p
              class="text-sm text-gray-500"
            >
              02 Jul 2024
            </p>
          </div>
        </div>
        <div
          class="mb-0"
        >
          <h3
            class="text-xl font-semibold mb-2"
          >
            Dietary Preferences
          </h3>
           
          <div
            class="flex flex-wrap gap-2"
          >
             
            <span
              class="bg-purple-100 text-purple-800 text-sm font-medium px-3 py-1 rounded-full"
            >
              Recipe_1_preference_1
               
            </span>
            <span
              class="bg-purple-100 text-purple-800 text-sm font-medium px-3 py-1 rounded-full"
            >
              Recipe_1_preference_2
               
            </span>
          </div>
        </div>
      </div>
      <div
        class="p-6"
      >
        <h3
          class="text-lg font-semibold mb-4"
        >
          Ask the AI Assistant
        </h3>
        <div
          class="mt-6 flex flex-col gap-4"
        >
          <div
            class="flex items-center gap-2"
          >
            <textarea
              class="flex-grow border rounded-lg px-4 py-2 text-sm h-[80px] resize-none overflow-y-auto"
              placeholder="Ask a question about this recipe..."
            />
            <button
              aria-label="Send"
              class="bg-blue-600 text-white px-4 py-2 rounded-lg text-sm hover:bg-blue-700 disabled:opacity-50"
              disabled=""
            >
              <svg
                aria-hidden="true"
                class="h-14 w-8"
                data-slot="icon"
                fill="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M3.478 2.404a.75.75 0 0 0-.926.941l2.432 7.905H13.5a.75.75 0 0 1 0 1.5H4.984l-2.432 7.905a.75.75 0 0 0 .926.94 60.519 60.519 0 0 0 18.445-8.986.75.75 0 0 0 0-1.218A60.517 60.517 0 0 0 3.478 2.404Z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
